# Borrower Reliability Analysis

**Customer:** Bank's credit department.

**Objective:** Test the hypothesis of whether a client's marital status and the number of children influence their ability to repay a loan on time.

**Input Data:** Client solvency statistics provided by the bank.

**Project Workflow:**

The analysis is conducted in 4 main stages:

1. Data Overview
2. Data Preprocessing
3. Hypothesis Testing
4. Conclusion

> **Outcome:**
The results of this study will inform the development of a credit scoring model — a specialized system used to evaluate a potential borrower's creditworthiness.

## Data Overview

In [1]:
# Importing the pandas library for data manipulation
import pandas as pd  

In [2]:
# Loading the dataset from the specified path
df = pd.read_csv('/datasets/data.csv')

In [3]:
# Displaying the first 20 rows of the dataset
df.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [4]:
# Displaying a summary of the dataset, including data types and non-null values
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [5]:
# Checking for missing values in each column and summing the total number of missing values
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [6]:
# Calculating the proportion of missing values in each columndf.isna().mean()
df.isna().mean()

children            0.000000
days_employed       0.100999
dob_years           0.000000
education           0.000000
education_id        0.000000
family_status       0.000000
family_status_id    0.000000
gender              0.000000
income_type         0.000000
debt                0.000000
total_income        0.100999
purpose             0.000000
dtype: float64

The dataset consists of 12 columns. According to the data documentation:

1. children — number of children in the family
2. days_employed — total employment duration in days
3. dob_years — client's age in years
4. education — client's education level
5. education_id — education level identifier
6. family_status — marital status
7. family_status_id — marital status identifier
8. gender — client's gender
9. income_type — type of employment
10. debt — whether the client has a loan repayment default
11. total_income — monthly income
12. purpose — loan purpose

In [7]:
# Iterating through each column in the DataFrame and displaying the unique value counts for each
for column in df:
    display(df[column].value_counts().unique) 

<bound method Series.unique of  0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64>

<bound method Series.unique of -327.685916     1
-1580.622577    1
-4122.460569    1
-2828.237691    1
-2636.090517    1
               ..
-7120.517564    1
-2146.884040    1
-881.454684     1
-794.666350     1
-3382.113891    1
Name: days_employed, Length: 19351, dtype: int64>

<bound method Series.unique of 35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64>

<bound method Series.unique of среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64>

<bound method Series.unique of 1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64>

<bound method Series.unique of женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64>

<bound method Series.unique of 0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64>

<bound method Series.unique of F      14236
M       7288
XNA        1
Name: gender, dtype: int64>

<bound method Series.unique of сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
студент                1
в декрете              1
Name: income_type, dtype: int64>

<bound method Series.unique of 0    19784
1     1741
Name: debt, dtype: int64>

<bound method Series.unique of 169846.427535    1
257737.077768    1
200508.675866    1
106196.235958    1
248730.171354    1
                ..
175057.266090    1
101516.604975    1
239154.168013    1
165009.733021    1
189255.286637    1
Name: total_income, Length: 19351, dtype: int64>

<bound method Series.unique of свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
покупка своего жилья                      620
строительство недвижимости                620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
зан

In [8]:
# Checking for explicit duplicates
df.duplicated().sum() #проверяем наличие явных дубликатов

54

In [9]:
# Summarizing statistics of numerical columns
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

### Intermediate Findings

1. The number of values in the columns differs, indicating missing data. Two columns, `days_employed` and `total_income`, have the same number of missing values, and the rows with missing data are identical.
2. The proportion of missing values in the columns is 10.1%. This percentage is not critical for hypothesis testing, especially for the `days_employed` column. Data related to marital status, loan defaults, and the number of children may be more relevant for the analysis.
3. Since both columns are related to the client's employment status (total work experience and monthly income), the missing data in these columns might be interrelated. The hypothesis for the missing data is that it may be due to a technical issue during data entry or extraction, as the missing values always correspond to the same rows.
4. Missing data will be imputed using the median. The median is preferred when there are outliers (extremely high or low values). Although the data review suggests these columns may not be critical for the analysis, the missing values will still be handled as per the technical task.
5. A check for unique values in some columns revealed the following anomalies:
    1. In the `children` column, the values -1 (47 occurrences) and 20 (76 occurrences) are found. These will be replaced with the median.
    2. The `days_employed` column contains negative values, which are invalid.
    3. The `dob_years` column contains 101 occurrences of the value 0. These will be replaced with the median.
    4. The `education` column contains inconsistent capitalization. Values will be standardized to a common format.
    5. In the `gender` column, the value XNA appears once. This will be excluded from the analysis.
    6. The `purpose` column has many implicit duplicates. Similar values will be grouped together.

## Data Preprocessing

### Filling Missing Values

In [10]:
# Calculating the median value of the 'children' column
df['children'].median()

0.0

In [11]:
# Replacing invalid values in the 'children' column
df['children'] = df['children'].replace(-1, 0)
df['children'] = df['children'].replace(20, 0)

In [12]:
# To fill missing values in the 'dob_years' column, the median will be calculated for each borrower group
df.groupby('income_type')['dob_years'].median()

income_type
безработный        38.0
в декрете          39.0
госслужащий        40.0
компаньон          39.0
пенсионер          60.0
предприниматель    42.5
сотрудник          39.0
студент            22.0
Name: dob_years, dtype: float64

In [13]:
# Since fillna() only works with NaN or None, it cannot be used for replacing values in 'dob_years'.
# Instead, loc is used to fill missing values based on the median of each 'income_type' group.
for t in df['income_type'].unique():
    df.loc[(df['income_type'] == t) & (df['dob_years'].isna()), 'dob_years'] = \
    df.loc[(df['income_type'] == t), 'dob_years'].median()

Since the `days_employed` column contains many outliers due to negative values, and this column is not critical for further analysis, we will only fill the missing values with the median. Since work experience is directly related to the borrower's age, we will create age groups and calculate the median for each group.

- Group: 19-25
- Group: 25-35
- Group: 35-45
- Group: 45-55
- Group: 55-65
- Group: 65-75

In [14]:
# Function to assign age groups based on 'dob_years'
def days_employed(row):  
    age = row['dob_years']

    if age <= 25:
        return 'Group 1'
    
    if age <= 35 and age > 25:
        return 'Group 2'
    
    if age <= 45 and age > 35:
        return 'Group 3'

    if age <= 55 and age > 45:
        return 'Group 4'
    
    if age <= 65 and age > 55:
        return 'Group 5'
    
    if age <= 75 and age > 65:
        return 'Group 6'

In [15]:
# Creating a new column 'age_group' with the results of the 'days_employed' function
df['age_group'] = df.apply(days_employed, axis=1) 

In [16]:
# Calculating the median of 'days_employed' for each age group
df.groupby('age_group')['days_employed'].median() 

age_group
Group 1      -812.631159
Group 2     -1337.959217
Group 3     -1834.869762
Group 4     -1672.165344
Group 5    344187.361114
Group 6    360948.880425
Name: days_employed, dtype: float64

In [17]:
# Removing negative values by converting 'days_employed' to absolute values
df['days_employed'] = df['days_employed'].abs() 

In [18]:
# Calculating the median 'days_employed' for each 'age_group'
df.groupby('age_group')['days_employed'].median()

age_group
Group 1       846.213341
Group 2      1350.995300
Group 3      1917.145342
Group 4      2886.495612
Group 5    344187.361114
Group 6    360948.880425
Name: days_employed, dtype: float64

> Negative values in this column are observed only in the two oldest age groups. There is a hypothesis that the format for recording work experience has changed.

In [19]:
# Checking the number of missing values to verify if the function worked correctly
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
age_group              0
dtype: int64

In [20]:
# Calculating the median 'total_income' for each 'income_type' group
df.groupby('income_type')['total_income'].median()

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

In [21]:
# Filling missing values in 'total_income' by replacing them with the median income for each 'income_type'
for t in df['income_type'].unique():
    df.loc[(df['income_type'] == t) & (df['total_income'].isna()), 'total_income'] = \
    df.loc[(df['income_type'] == t), 'total_income'].median()

In [22]:
# Checking the number of missing values after filling
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
age_group              0
dtype: int64

In [23]:
# Filling missing 'days_employed' values by replacing them with the median for each 'age_group'
for t in df['age_group'].unique():
    df.loc[(df['age_group'] == t) & (df['days_employed'].isna()), 'days_employed'] = \
    df.loc[(df['age_group'] == t), 'days_employed'].median()

In [24]:
# Checking the number of missing values after filling
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
age_group           0
dtype: int64

In [25]:
# Excluding 'XNA' from the 'gender' column and resetting the index
df = df[df['gender'] != 'XNA'].reset_index(drop=True)

### Data Anomaly Detection and Correction

In [26]:
# Checking if the applied methods worked correctly
df['children'].value_counts()

0    14271
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

In [27]:
# Checking the unique values in the 'gender' column
df['gender'].unique()

array(['F', 'M'], dtype=object)

In [28]:
# Displaying information about the DataFrame, including column types and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21524 entries, 0 to 21523
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21524 non-null  int64  
 1   days_employed     21524 non-null  float64
 2   dob_years         21524 non-null  float64
 3   education         21524 non-null  object 
 4   education_id      21524 non-null  int64  
 5   family_status     21524 non-null  object 
 6   family_status_id  21524 non-null  int64  
 7   gender            21524 non-null  object 
 8   income_type       21524 non-null  object 
 9   debt              21524 non-null  int64  
 10  total_income      21524 non-null  float64
 11  purpose           21524 non-null  object 
 12  age_group         21524 non-null  object 
dtypes: float64(3), int64(4), object(6)
memory usage: 2.1+ MB


In [29]:
# Generating summary statistics for numerical columns in the DataFrame
df.describe()

children  days_employed     dob_years  education_id  \
count  21524.000000   21524.000000  21524.000000  21524.000000   
mean       0.470498   67749.324333     43.294276      0.817181   
std        0.750544  139249.097819     12.574188      0.548092   
min        0.000000      24.141633      0.000000      0.000000   
25%        0.000000    1005.821788     33.000000      1.000000   
50%        0.000000    2143.533786     42.000000      1.000000   
75%        1.000000    5430.905847     53.000000      1.000000   
max        5.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21524.000000  21524.000000  2.152400e+04  
mean           0.972542      0.080886  1.652235e+05  
std            1.420357      0.272667  9.804559e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077960e+05  
50%            0.000000      0.000000  1.425944e+05  
75%            1.000000      0.000000  1.955452e+05  
max            4.000000      1.000000  2.265604e+06

### Data Type Conversion

In [30]:
# Converting 'total_income' from float to integer data type
df['total_income'] = df['total_income'].astype('int')

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21524 entries, 0 to 21523
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21524 non-null  int64  
 1   days_employed     21524 non-null  float64
 2   dob_years         21524 non-null  float64
 3   education         21524 non-null  object 
 4   education_id      21524 non-null  int64  
 5   family_status     21524 non-null  object 
 6   family_status_id  21524 non-null  int64  
 7   gender            21524 non-null  object 
 8   income_type       21524 non-null  object 
 9   debt              21524 non-null  int64  
 10  total_income      21524 non-null  int64  
 11  purpose           21524 non-null  object 
 12  age_group         21524 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


### Removing Duplicates

Removing explicit duplicates. We identified 54 duplicates during the [data overview](#intro) step.

In [32]:
# Removing duplicates and resetting the index to avoid gaps
df = df.drop_duplicates().reset_index()

In [33]:
# Checking the number of duplicate rows in the DataFrame
df.duplicated().sum() 

0

Handling duplicates in the `education` column caused by inconsistent formatting.
Creating a function to convert all implicit duplicates to snake_case.

In [34]:
# Viewing unique values in the 'education' column after sorting
df['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

In [35]:
# Converting all values in the 'education' column to lowercase
df['education'] = df['education'].str.lower()

In [36]:
# Checking the unique values in the 'education' column after the corrections
df['education'].sort_values().unique() #проверяем исправления

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

### Creating Additional DataFrames and Decomposing the Original DataFrame

In [37]:
# Creating a new DataFrame 'df_education' with unique 'education' and 'education_id' pairs
df_education = df[['education','education_id']] 

In [38]:
# Removing duplicates
df_education = df_education.drop_duplicates().reset_index(drop=True) 
df_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [39]:
# Creating a new DataFrame 'df_family_status' with unique 'family_status' and 'family_status_id' pairs
df_family_status = df[['family_status','family_status_id']] 
df_family_status = df_family_status.drop_duplicates().reset_index(drop=True)
df_family_status

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [40]:
# Removing 'education' and 'family_status' columns from the original DataFrame
df = df.drop('education', axis=1) #удаляем из исходного датафрейма столбцы education и family_status

In [41]:
# Dropping the 'family_status' column from the DataFrame
df = df.drop('family_status', axis=1)

In [42]:
df.head()

index  children  days_employed  dob_years  education_id  family_status_id  \
0      0         1    8437.673028       42.0             0                 0   
1      1         1    4024.803754       36.0             1                 0   
2      2         0    5623.422610       33.0             1                 0   
3      3         3    4124.747207       32.0             1                 0   
4      4         0  340266.072047       53.0             1                 1   

  gender income_type  debt  total_income                     purpose age_group  
0      F   сотрудник     0        253875               покупка жилья   Group 3  
1      F   сотрудник     0        112080     приобретение автомобиля   Group 3  
2      M   сотрудник     0        145885               покупка жилья   Group 2  
3      M   сотрудник     0        267628  дополнительное образование   Group 2  
4      F   пенсионер     0        158616             сыграть свадьбу   Group 4

### Income Categorization

Creating a new column `total_income_category` with income categories:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 and above — 'A'.

In [43]:
# Function to assign income categories
def income_category(row):
    
    income = row['total_income']

    if income <= 30000:
        return 'E'
    
    if income <= 50000 and income > 30001:
        return 'D'
    
    if income <= 200000 and income > 50001:
        return 'C'

    if income <= 1000000 and income > 200001:
        return 'B'
    
    if income > 1000001:
        return 'A'

In [44]:
# Apply the function to each row
df['total_income_category'] = df.apply(income_category, axis=1)
df.head()

index  children  days_employed  dob_years  education_id  family_status_id  \
0      0         1    8437.673028       42.0             0                 0   
1      1         1    4024.803754       36.0             1                 0   
2      2         0    5623.422610       33.0             1                 0   
3      3         3    4124.747207       32.0             1                 0   
4      4         0  340266.072047       53.0             1                 1   

  gender income_type  debt  total_income                     purpose  \
0      F   сотрудник     0        253875               покупка жилья   
1      F   сотрудник     0        112080     приобретение автомобиля   
2      M   сотрудник     0        145885               покупка жилья   
3      M   сотрудник     0        267628  дополнительное образование   
4      F   пенсионер     0        158616             сыграть свадьбу   

  age_group total_income_category  
0   Group 3                     B  
1   Group 3                     C  
2   Group 2                     C  
3   Group 2                     B  
4   Group 4                     C

### Categorizing Loan Purposes

We will create a function that, based on the data in the `purpose` column, will generate a new column called `purpose_category` with the following categories:

* 'vehicle-related transactions'
* 'real estate transactions'
* 'wedding-related expenses'
* 'education expenses'.

In [45]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [50]:
# Function to categorize loan purposes based on the 'purpose' column
def purpose_category(row):
    
    purpose = row['purpose']
    
    if 'авто'in purpose:
        return 'vehicle-related transactions'
    
    if 'жил' in purpose or 'недвиж' in purpose:
        return 'real estate transactions'
    
    if 'свадьб' in purpose:
        return 'wedding-related expenses'
    
    if 'обр' in purpose:
        return 'education expenses' 

In [53]:
# Create a new column 'purpose_category' based on the 'purpose' column using the function purpose_category
df['purpose_category'] = df.apply(purpose_category, axis=1) 

In [54]:
# Check unique values in the 'purpose_category' column
df['purpose_category'].unique()

array(['real estate transactions', 'vehicle-related transactions',
       'education expenses', 'wedding-related expenses'], dtype=object)

### Intermediate Findings

At this stage, the data was modified as follows:

1. Missing values in the `dob_years` and `total_income` columns were replaced with median values.
2. Anomalies in the `children` and `dob_years` columns were processed and replaced with median values.
3. Negative values in the `days_employed` column were converted to positive values.
4. Borrowers were grouped by age and income, with the results stored in the `age_group` and `income_category` columns.
5. 54 explicit duplicates and 1 row with an undefined gender value were removed.
6. Values in the `purpose` column were standardized.
7. The `education` and `family_status` columns were removed from the dataframe.

## Answers to the Questions

### Question 1: Is there a relationship between the number of children and timely loan repayment?

In [55]:
# Creates a pivot table to analyze the relationship between the number of children and debt repayment
df.pivot_table(index=['children'], values='debt', aggfunc=['count','mean'])

count      mean
           debt      debt
children                 
0         14229  0.075339
1          4809  0.092327
2          2052  0.094542
3           330  0.081818
4            41  0.097561
5             9  0.000000

#### Conclusion 1

The samples by the number of children vary significantly, ranging from thousands to just a few, making it difficult to draw conclusions about the relationship based on the current data.

However, if we consider only the comparable categories—borrowers without children and borrowers with 1 or 2 children—we can observe a trend: **the more children in the family, the more likely borrowers are to have outstanding debts.**


### Question 2: Is there a relationship between marital status and timely loan repayment?

In [57]:
# Create a pivot table to analyze the relationship between marital status and loan repayment.
df.pivot_table(index=['family_status_id'], values='debt', aggfunc=['count','mean'])

count      mean
                   debt      debt
family_status_id                 
0                 12344  0.075421
1                  4162  0.093224
2                   959  0.065693
3                  1195  0.071130
4                  2810  0.097509

#### Conclusion 2

On the current sample, we observe the following relationship between marital status and loan repayment:

From the group with the least tendency towards loan default to the group with the highest tendency towards defaults:

- Widowed (вдовец / вдова)
- Divorced (в разводе)
- Married (женат / замужем)
- Cohabiting (гражданский брак)
- Single (не женат / не замужем)

### Question 3: Is there a relationship between income level and timely loan repayment?

In [59]:
#Creating a pivot table to analyze the relationship between income category and loan repayment. 
df.pivot_table(index=['total_income_category'], values='debt', aggfunc=['count','mean'])

count      mean
                        debt      debt
total_income_category                 
A                         25  0.080000
B                       5040  0.070635
C                      16032  0.084830
D                        350  0.060000
E                         22  0.090909

#### Conclusion 3

The situation with the small sample size repeats itself—there are far fewer borrowers in category A, but even so, it is difficult to identify a clear relationship in the data. Borrowers in category A and category E have almost identical debt ratios, despite having incomes differing by nearly a million.

Taking into account the well-known fact that people with lower incomes tend to repay their debts more diligently than those with higher incomes, we can conclude that **this effect is evident in our data, even if it is based on a relatively small sample size.** 

### Question 4: How do different loan purposes affect timely repayment?

In [60]:
# Creates a pivot table to analyze the relationship between loan purposes and timely repayment. 
df.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count','mean'])

count      mean
                               debt      debt
purpose_category                             
education expenses             4014  0.092177
real estate transactions      10813  0.072320
vehicle-related transactions   4308  0.093547
wedding-related expenses       2335  0.079657

#### Conclusion 4

There is a clear dependency observed: borrowers in the categories of `real estate transactions` and `wedding-related expenses` tend to repay their loans on time more often, while borrowers in the categories of `vehicle-related transactions` and `education expenses` are more likely to have loan arrears.

## General Conclusion

During the course of this work, critical anomalies in the provided data were addressed, explicit and implicit duplicates were removed, and missing values were filled.

**Here are the key conclusions:**

1. **Larger Family Size and Loan Arrears:** The more children a family has, the more likely borrowers are to experience loan arrears. It is important to note that the samples for the number of children are not proportional, which makes this finding less conclusive.

2. **Widows/Widowers vs. Single Borrowers:** Borrowers in the "widow/widower" group tend to repay their loans on time more frequently than those in the "never married" group, suggesting that marital status may play a role in loan repayment behavior.

3. **Income and Loan Repayment:** It is challenging to establish a clear and logical relationship between income level and timely loan repayment. Interestingly, borrowers in both the highest income category (A) and the lowest income category (E) show almost identical default rates, despite a nearly one million difference in their incomes. This might indicate that income alone does not sufficiently explain repayment behavior, and other factors, such as financial discipline or obligations, may play a role.

4. **Purpose of Loan and Repayment:** Borrowers in the "real estate transactions" and "wedding-related expenses" categories are more likely to repay their loans on time. On the other hand, those in the "vehicle-related transactions" and "education expenses" categories tend to have more loan arrears. This suggests that the purpose of the loan might influence the borrower's likelihood of meeting repayment deadlines, with non-essential purchases (like cars) and investments in education showing higher rates of arrears.